### PR TITLE
Support inputs to ORTModule forward method that require gradient

### PR DESCRIPTION
**Pull request by baijumeswani**
**Description**: Support for inputs to the ```ORTModule.forward()``` method that ```require_grad```

This pull request addresses the following:
- Failure in trying to ```copy.deepcopy(inputs)``` for input tensors that had the ```grad_fn``` attribute.
- Register the inputs that require gradient computation with the ```mpdule_gradient_graph_builder```.
- Populate the results in the ```backward()``` so that it has the output for the inputs that require grad.
- Re-initialize the graph in case the inputs requiring grad change from one forward call to the next.
- Unit tests for the four points above.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -391,19 +423,12 @@ def _get_forward_graph(module, *inputs, **kwargs):
...
-        input_names = []
-        dynamic_axes = {}
-        for input_idx, name in enumerate(all_input_names):
-            if input_idx < len(inputs) and inputs[input_idx] is not None:
-                input_names.append(name)
-                dynamic_axes[name] = {}
-                for dim_idx in range(len(inputs[input_idx].shape)):
-                    dynamic_axes[name].update({dim_idx : f'input{input_idx}_dim{dim_idx}'})
+        # Inputs may contain tensors that have attributes preventing their deepcopy (example grad_fn).
+        # Deepcopy only the data component of the input tensors for export.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:427]: How about making this comment a little bit more "scary" by prepending IMPORTANT: or WARNING: or NOTE: or something like that? This change may have side effects, so better give a clue for future us

File: orttraining/orttraining/python/training/ortmodule.py [hunk 2]
```diff
@@ -379,9 +413,7 @@ def _convert_backward_input_list_to_dict(self, *inputs):
 
         return result
 
-
-    @staticmethod
-    def _get_forward_graph(module, *inputs, **kwargs):
+    def _get_forward_graph(self, module, input_names, dynamic_axes, *inputs, **kwargs):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:416]: What is the rationale in making it not static and still pass `module` as argument? This parameter could be suppressed and internally `self._original_module`, right?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -239,3 +239,37 @@ def test_model_to_device_and_back_to_original(original_device, to_device):
...
+    model(x)
+    assert(model._input_names_require_grad == ['input1'])
+
+def test_input_requires_grad_backward_creates_input_grad():
+    device = 'cuda'
+
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
+    model = ORTModule(model)
+    x = torch.randn(N, D_in, device=device, requires_grad=True)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:259]:
```suggestion
    x = torch.randn(N, D_in, device=device, requires_grad=True)
    assert x.grad is None
```

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -239,3 +239,37 @@ def test_model_to_device_and_back_to_original(original_device, to_device):
     assert model._device == torch.device(original_device+':0')
     for _, parameter_value in model.named_parameters():
         assert parameter_value.device.type == original_device
+
+def test_input_requires_grad_saved():
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:243]: This test is quiet simple and could be merged with the one below

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -239,3 +239,37 @@ def test_model_to_device_and_back_to_original(original_device, to_device):
...
+
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
+    model = ORTModule(model)
+    x = torch.randn(N, D_in, device=device, requires_grad=True) + 1
+    model(x)
+    assert(model._input_names_require_grad == ['input1'])
+
+def test_input_requires_grad_backward_creates_input_grad():
+    device = 'cuda'
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:254]:
```suggestion
@pytest.mark.parametrize("device", ['cuda', 'cpu'])
def test_input_requires_grad_backward_creates_input_grad(device):
```

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 6]
```diff
@@ -239,3 +239,37 @@ def test_model_to_device_and_back_to_original(original_device, to_device):
...
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
+    model = ORTModule(model)
+    x = torch.randn(N, D_in, device=device, requires_grad=True)
+    prediction = model(x)
+    s = prediction.sum()
+    s.backward()
+    assert x.grad is not None
+
+def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder():
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:265]:
```suggestion
@pytest.mark.parametrize("device", ['cuda', 'cpu'])
def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder(device):
```

**Review by thiagocrepaldi — COMMENTED** [review]
Very nice. I've made some minor comments

File: orttraining/orttraining/python/training/ortmodule.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:416]: you're right. I initially wanted to keep it static and then realized it needn't be static. Missed removing the module argument. Will remove that.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 4 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:243]: I actually require this test. Here I want to test that a non leaf tensor can be correctly deepcopied (which is why i use ```x = tensor(...)+1```).
And in the below test, I cannot use ```x = tensor(...)+1``` because I need to ```assert x.grad is not None```.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 7]
```diff
@@ -180,21 +203,22 @@ def forward(self, *inputs, **kwargs):
...
+            input_names, dynamic_axes, self._input_names_require_grad = \
+                _extract_input_information(self._original_module, *inputs, **kwargs)
+            self._onnx_training = self._get_forward_graph(self._original_module, input_names, dynamic_axes, *inputs, **kwargs)
 
-            # Build full training graph and split in forward/backward
-            grad_builder_config = C.ModuleGradientGraphBuilderConfiguration()
-            grad_builder_config.initializer_names_to_train = initializer_names
-            grad_builder_config.input_names_require_grad = []
-            self._module_gradient_graph_builder = C.ModuleGradientGraphBuilder()
-            self._module_gradient_graph_builder.initialize(self._onnx_training.SerializeToString(), grad_builder_config)
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py]: I adjusted code to account for the changes in that PR.

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thanks.
